pkg/controllers/node/termination/suite_test.go

```diff
@@ -706,10 +716,132 @@ var _ = Describe("Termination", func() {
+			ExpectNodeExists(ctx, env.Client, node.Name)
+
+			Expect(env.Client.Get(ctx, types.NamespacedName{Name: node.Name}, node)).To(Succeed())
+			Expect(node.Spec.Taints).ToNot(ContainElement(v1beta1.DisruptionNonGracefulShutdown))
+		})
+		It("should not taint a node that has not yet exceeded it's terminationGracePeriod", func() {
+			minAvailable := intstr.FromInt(1)
+			labelSelector := map[string]string{test.RandomName(): test.RandomName()}
+			pdb := test.PodDisruptionBudget(test.PDBOptions{
+				Labels: labelSelector,
+				// Don't let any pod evict
+				MinAvailable: &minAvailable,
+			})
+			podNoEvict := test.Pod(test.PodOptions{
+				NodeName: node.Name,
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:          labelSelector,
+					OwnerReferences: defaultOwnerRefs,
+				},
+				Phase: v1.PodRunning,
+			})
+
+			node.ObjectMeta.Annotations = map[string]string{
+				v1beta1.NodeExpirationTimeAnnotationKey: time.Now().Add(time.Minute * 5).Format(time.RFC3339),
+			}
+			nodePool.Spec.Disruption.TerminationGracePeriod = &metav1.Duration{Duration: time.Second * 300}
+			ExpectApplied(ctx, env.Client, node, nodeClaim, nodePool, podNoEvict, pdb)
+
+			Expect(env.Client.Delete(ctx, node)).To(Succeed())
+			ExpectObjectReconciled(ctx, env.Client, terminationController, node)
+			ExpectSingletonReconciled(ctx, queue)
+			ExpectNodeExists(ctx, env.Client, node.Name)
+			ExpectPodExists(ctx, env.Client, podNoEvict.Name, podNoEvict.Namespace)
+			Expect(node.Spec.Taints).ToNot(ContainElement(v1beta1.DisruptionNonGracefulShutdown))
+		})
+		It("should taint a node that has exceeded it's terminationGracePeriod", func() {
+			minAvailable := intstr.FromInt(1)
+			labelSelector := map[string]string{test.RandomName(): test.RandomName()}
+			pdb := test.PodDisruptionBudget(test.PDBOptions{
+				Labels: labelSelector,
+				// Don't let any pod evict
+				MinAvailable: &minAvailable,
+			})
+			podNoEvict := test.Pod(test.PodOptions{
+				NodeName: node.Name,
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:          labelSelector,
+					OwnerReferences: defaultOwnerRefs,
+				},
+				Phase: v1.PodRunning,
+			})
+
+			node.ObjectMeta.Annotations = map[string]string{
+				v1beta1.NodeExpirationTimeAnnotationKey: time.Now().Add(time.Second * -10).Format(time.RFC3339),
+			}
+			nodePool.Spec.Disruption.TerminationGracePeriod = &metav1.Duration{Duration: time.Minute * 1}
+			ExpectApplied(ctx, env.Client, node, nodeClaim, nodePool, podNoEvict, pdb)
+
+			Expect(env.Client.Delete(ctx, node)).To(Succeed())
+			ExpectObjectReconciled(ctx, env.Client, terminationController, node)
+			ExpectSingletonReconciled(ctx, queue)
+			ExpectNodeExists(ctx, env.Client, node.Name)
+
+			Expect(env.Client.Get(ctx, types.NamespacedName{Name: node.Name}, node)).To(Succeed())
+			Expect(node.Spec.Taints).To(ContainElement(v1beta1.DisruptionNonGracefulShutdown))
+		})
+		It("should preemptively delete pods to satisfy their terminationGracePeriodSeconds", func() {
```

Comment:
Could we add a check to this test to ensure that the pod will delete at 60 seconds rather than 300?

Reply:
This pod is deleted at 0 seconds, because it has a TGPS of 300 while the node has a TGP of 60. (60 - 300) is 240s before the node begins draining, so the pod is deleted immediately to give it the maximum amount of time possible to delete gracefully prior to the node's TGP expiring.

The next test does cover the case where the pod should be deleted 60s into the node draining period (maximum 300s). It verifies the pod still exists after the node is deleted, advances the fake clock by 90s, then confirms the pod has been deleted.